ios/NotificationServiceExtension/NotificationService.swift

```diff
@@ -88,10 +88,61 @@ class NotificationService: UNNotificationServiceExtension {
       // TODO: Need to understand to who this notification was delivered to.
       let accounts = Table("accounts")
       let tokens = Table("tokens")
+      let locks = Table("locks")
 
+      let operation = SQLite.Expression<String>("operation")
+      let lockedAt = SQLite.Expression<Int>("locked_at")
       let userId = SQLite.Expression<String>("user_id")
       let credentials = SQLite.Expression<String>("credentials")
 
+      var acquired = false
+
+      while !acquired {
+        try? db.transaction {
+          let query =
+            try locks
+            .select(lockedAt)
+            .where(operation == "refreshSession")
+            .limit(1)
```

Comment:
@SleepySquash could we extract this locking logic to separate functions/class? They're just too noisy here, in the `acknowledgeDelivery` function.